pkg/lightning/backend/kv/session.go

```diff
@@ -438,34 +438,75 @@ func (*session) GetStmtStats() *stmtstats.StatementStats {
 
 // Session is used to provide context for lightning.
 type Session struct {
-	sctx *session
+	txn     transaction
+	exprCtx *litExprContext
+	tblCtx  *litTableMutateContext
 }
 
 // NewSession creates a new Session.
-func NewSession(options *encode.SessionOptions, logger log.Logger) *Session {
-	return &Session{
-		sctx: newSession(options, logger),
+func NewSession(options *encode.SessionOptions, logger log.Logger) (*Session, error) {
+	sysVars := make(map[string]string, len(options.SysVars))
+	if options.SysVars != nil {
+		// This sessVars is only used to do validations.
+		sessVars := variable.NewSessionVars(nil)
+		// To keep compatible with the old versions, we should to skip errors caused by illegal system variables.
+		for k, v := range options.SysVars {
+			// since 6.3(current master) tidb checks whether we can set a system variable
+			// lc_time_names is a read-only variable for now, but might be implemented later,
+			// so we not remove it from defaultImportantVariables and check it in below way.
+			if sv := variable.GetSysVar(k); sv == nil {
+				logger.DPanic("unknown system var", zap.String("key", k))
+				continue
+			} else if sv.ReadOnly {
+				logger.Debug("skip read-only variable", zap.String("key", k))
+				continue
+			}
+			if err := sessVars.SetSystemVar(k, v); err != nil {
+				logger.DPanic("new session: failed to set system var",
+					log.ShortError(err),
+					zap.String("key", k))
+				continue
+			}
+			sysVars[k] = v
+		}
+	}
+
+	exprCtx, err := newLitExprContext(options.SQLMode, sysVars, options.Timestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	tblCtx, err := newLitTableMutateContext(exprCtx, sysVars)
+	if err != nil {
+		return nil, err
 	}
+
+	s := &Session{
+		exprCtx: exprCtx,
+		tblCtx:  tblCtx,
+	}
+	s.txn.kvPairs = &Pairs{}
+	return s, nil
 }
 
 // GetExprCtx returns the expression context
-func (s *Session) GetExprCtx() expression.BuildContext {
-	return s.sctx.GetExprCtx()
+func (s *Session) GetExprCtx() exprctx.ExprContext {
+	return s.exprCtx
 }
 
 // Txn returns the internal txn.
 func (s *Session) Txn() kv.Transaction {
-	return &s.sctx.txn
+	return &s.txn
 }
 
 // GetTableCtx returns the table MutateContext.
 func (s *Session) GetTableCtx() tbctx.MutateContext {
-	return s.sctx.tblctx
+	return s.tblCtx
 }
 
 // TakeKvPairs returns the current Pairs and resets the buffer.
 func (s *Session) TakeKvPairs() *Pairs {
```

Comment:
lightning use `TakeKvPairs` to get encoded KV paris. Can we use `GetMutateBuffers` in future?

Reply:
Currently, `GetMutateBuffers` is only designed for implementation in `tables` package so it does not provide general object pools. However, if you want to reuse the memory, the simplest way is to maintain a memory buffer in `Session` itself.